File: ios/Goo/Layout.swift
import Foundation
import Yaml

public enum LayoutType: String {
    case relative = "relative"
    case vertical = "vertical"
    case horizontal = "horizontal"
    case grid = "grid"
}

public class Layout {
    let type: LayoutType
    let padding: Inset
    let spacing: Vector
    let alignment: Alignment
    let cellSize: Vector?
    let distribute: Vector
    let fit: Vector
    
    let weight: Double
    let ignore: Bool
    
    init(_ yaml: Yaml) {
        type = yaml["layout"].string.flatMap { LayoutType(rawValue: $0) } ?? .vertical
        
        let paddingProp = yaml["padding"]
        padding = paddingProp.double.map { Inset($0) } ?? Inset(paddingProp) ?? Inset.zero
        
        let spacingProp = yaml["spacing"]
        spacing = Vector.one * spacingProp.double ?? Vector(spacingProp) ?? Vector.zero
        
        alignment = yaml["alignment"].string.flatMap { Alignment(rawValue: $0) } ?? .topLeft
        
        let cellSizeProp = yaml["cell-size"]
        cellSize = Vector.one * cellSizeProp.double ?? Vector(cellSizeProp)
        
        distribute = Vector.direction(yaml["distribute"])
        
        fit = Vector.direction(yaml["fit"])
        
        weight = yaml["weight"].double ?? 1
        
        ignore = yaml["ignore-layout"].bool ?? false
    }
    
}

File: android/app/src/main/java/org/haralovich/goo/Layout.kt
package org.haralovich.goo

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.NullNode

enum class LayoutType {
    RELATIVE,
    VERTICAL,
    HORIZONTAL,
    GRID;

    companion object {

        fun parse(json: JsonNode): LayoutType? {
            return when (json.textOption) {
                "relative" -> RELATIVE
                "horizontal" -> HORIZONTAL
                "grid" -> GRID
                "vertical" -> VERTICAL
                else -> null
            }
        }

    }

}

class Layout(
    val type: LayoutType,
    val padding: Inset,
    val spacing: Vector,
    val alignment: Alignment,
    val cellSize: Vector?,
    val distribute: Vector,
    val fit: Vector,
    val weight: Float,
    val ignore: Boolean
) {

    companion object {

        fun parse(json: JsonNode): Layout {
            val type = LayoutType.parse(json.path("type")) ?: LayoutType.VERTICAL

            val paddingProp = json.path("padding")
            val padding = if (paddingProp.isFloat) {
                Inset(paddingProp.floatValue())
            } else {
                Inset.parse(paddingProp) ?: Inset.zero
            }

            val spacingProp = json.path("spacing")
            val spacing = if (spacingProp.isFloat) {
                Vector.one * spacingProp.floatValue()
            } else {
                Vector.parse(spacingProp) ?: Vector.zero
            }

            val alignment = Alignment.parse(json.path("alignment")) ?: Alignment.TOP_LEFT

            val cellSizeProp = json.path("cell-size")
            val cellSize = Vector.one * cellSizeProp.floatOption ?: Vector.parse(cellSizeProp)

            val distribute = Vector.direction(json.path("distribute"))

            val fit = Vector.direction(json.path("fit"))

            val weight = json.path("weight").floatOption ?: 1f

            val ignore = json.path("ignore-layout").booleanValue()

            return Layout(
                type = type,
                padding = padding,
                spacing = spacing,
                alignment = alignment,
                cellSize = cellSize,
                distribute = distribute,
                fit = fit,
                weight = weight,
                ignore = ignore
            )
        }

    }

}
